### PR TITLE
Support for named listeners v2

### PR DESCRIPTION
**Pull request by mhowlett**

### CODE REVIEW

File: core/src/main/java/io/confluent/rest/Application.java [hunk 1]
```diff
@@ -370,6 +386,8 @@ public Handler configureWebSocketHandler() throws ServletException {
   }
 
   // This is copied from the old MAP implementation from cp ConfigDef.Type.MAP
+  // TODO: This method should be removed in favor of RestConfig.getMap(). Note
+  // that it is no longer used by projects related to kafka-rest.
```
Comment (mhowlett) [inline core/src/main/java/io/confluent/rest/Application.java:390]: Note: I plan to make changes to Schema Registry following this set of PRs being merged to clean up Application.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 2]
```diff
@@ -399,23 +429,28 @@ private void configureConnectors(SslContextFactory sslContextFactory) {
...
-                                       URI listener,
+                                       NamedURI listener,
                                        boolean http2Enabled) {
     NetworkTrafficServerConnector connector;
 
     if (http2Enabled) {
-      log.info("Adding listener with HTTP/2: " + listener.toString());
-      if (listener.getScheme().equals("http")) {
+      log.info(
+          "Adding listener " + (listener.name == null ? "" : "'" + listener.name + "'")
```
Comment (rigelbm) [inline core/src/main/java/io/confluent/rest/ApplicationServer.java:451]: This should be in the NameURI toString().

**Review by rigelbm — APPROVED** [review]
The overall design looks perfect!

All comments are Java nits. Thank and LGTM!!!

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 3]
```diff
@@ -107,76 +107,103 @@ public ApplicationServer(T config, ThreadPool threadPool) {
     configureConnectors(sslContextFactory);
   }
 
+  static class NamedURI {
+    public final URI uri;
+    public final String name;
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:112]: i was forgetting idiomatic Java is maximal verbosity.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 4]
```diff
@@ -107,76 +107,103 @@ public ApplicationServer(T config, ThreadPool threadPool) {
...
+
+    public NamedURI(URI uri, String name) {
+      this.uri = uri;
+      this.name = name;
+    }
+  }
+
+  static NamedURI constructNamedURI(
+          String namedListenerMaybe,
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:121]: fair enough, `listener` is better.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 5]
```diff
@@ -107,76 +107,103 @@ public ApplicationServer(T config, ThreadPool threadPool) {
...
+      throw new ConfigException(
+          "Listener '" + namedListenerMaybe + "' is not a valid URI.");
+    }
+    if (uri.getPort() == -1) {
+      throw new ConfigException(
+          "Listener '" + namedListenerMaybe + "' must specify a port.");
+    }
+    if (!supportedSchemes.contains(uri.getScheme())) {
+      String uriName = uri.getScheme().toLowerCase();
+      String protocolMaybe = listenerProtocolMap.get(uriName);
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:137]: refactor hangover

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 6]
```diff
@@ -107,76 +107,103 @@ public ApplicationServer(T config, ThreadPool threadPool) {
...
+      if (protocolMaybe == null) {
+        throw new ConfigException(
+            "Listener '" + uri + "' has an unsupported scheme '" + uri.getScheme() + "'");
+      }
+      try {
+        uri = UriBuilder.fromUri(namedListenerMaybe).scheme(protocolMaybe).build();
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:143]: i think you're more pedantic than me

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 7]
```diff
@@ -107,76 +107,103 @@ public ApplicationServer(T config, ThreadPool threadPool) {
...
+
+    for (String listener : listeners) {
+      NamedURI namedURI = constructNamedURI(listener, listenerProtocolMap, supportedSchemes);
+      if (namedURI.name != null) {   // Multiple unnamed URIs are allowed.
+        for (NamedURI existingNamedURI : namedURIs) {
+          if (existingNamedURI.name != null && existingNamedURI.name.equals(namedURI.name)) {
+            throw new ConfigException(
+              "More than one listener with name '" + namedURI.name + "' was defined.");
+          }
+        }
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:187]: the 100 character line limit style check is really shitting me here.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java (hunk 2 above)
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:451]: well pointed out

File: core/src/main/java/io/confluent/rest/ApplicationServer.java [hunk 8]
```diff
@@ -107,76 +107,103 @@ public ApplicationServer(T config, ThreadPool threadPool) {
     configureConnectors(sslContextFactory);
   }
 
+  static class NamedURI {
+    public final URI uri;
+    public final String name;
+
+    public NamedURI(URI uri, String name) {
```
[the remark this answers is not in the document]
Comment (mhowlett) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:114]: unfortunately this is used in `Application` as well, so it needs to be public until we get rid of that code.

**Comment by mhowlett** [comment]
retest this please

**Pull request merged**